[Docs] translate logging design (#719)
open-mmlab/mmengine: docs/en/design/logging.md

# Logging

## Overview

> [Runner](./runner.md) produces a great amount of logs during execution. These logs include but not limited to dataset information, model initialization, learning rates and losses. In order to make these logs easily accessed by users, we designed [MessageHub](mmengine.logging.MessageHub), [HistoryBuffer](mmengine.logging.HistoryBuffer), [LogProcessor](mmengine.runner.LogProcessor) and [MMLogger](mmengine.logging.MMLogger) in MMEngine, which enables:

Reviewer: ```suggestion
[Runner](./runner.md) produces amounts of logs during execution. These logs include dataset information, model initialization, learning rates, losses, etc. In order to make these logs easily accessed by users, MMEngine designs [MessageHub](mmengine.logging.MessageHub), [HistoryBuffer](mmengine.logging.HistoryBuffer), [LogProcessor](mmengine.runner.LogProcessor) and [MMLogger](mmengine.logging.MMLogger), which enable:
```

> - Configuring statistical methods in config files. For example, you can configure to count losses by averaging across the whole training process instead of by smoothing in a sliding window manner
> - Querying training states (iterations, epochs, etc.) anywhere in any module
> - Switching the log saving method between multi-process and master process only

Reviewer: ```suggestion
- Configure statistical methods in config files. For example, losses can be globally averaged or smoothed by a sliding window.
- Query training states (iterations, epochs, etc.) in any module
- Configure whether save the multi-process log or not during distributed training.
```

![image](https://user-images.githubusercontent.com/57566630/163441489-47999f3a-3259-44ab-949c-77a8a599faa5.png)

> HistoryBuffer encapsules losses, learning rates and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../tutorials/logging.md) to familiarize yourself with some basic use cases.

Reviewer 2: ```suggestion
HistoryBuffer encapsulates losses, learning rates, and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../advanced_tutorials/logging.md) to familiarize yourself with some basic use cases.
```

> HistoryBuffer encapsulates losses, learning rates, and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../advanced_tutorials/logging.md) to familiarize yourself with some basic use cases.

Reviewer: ```suggestion
Each scalar (losses, learning rates, etc.) during training is encapsulated by HistoryBuffer, managed by MessageHub in key-value pairs, formatted by LogProcessor and then exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, statistical methods of these scalars can be configured through the LogProcessor without understanding the data flow.**  Before diving into the design of the logging system, please read through [logging tutorial](../advanced_tutorials/logging.md) first for familiarizing basic use cases.
```

## HistoryBuffer

> MMEngine provides HistoryBuffer class to record the trajectory of training. Scalars such as losses, learning rates and iteration time are saved to corresponding HistoryBuffer as an element in an array. In most cases, HistoryBuffer is used as an internal class. It works with [MessageHub](mmengine.logging.MessageHub), LoggerHook and [LogProcessor](mmengine.runner.LogProcessor) to make training log configurable.
> 
> However, HistoryBuffer can also be used alone, which enables users to manage their training logs and do various statistics in an easy manner. Therefore, we will first introduce the usage of HistoryBuffer in the following section. The association between HistoryBuffer and MessageHub will be introduced later in MessageHub section.

Reviewer: ```suggestion
`HistoryBuffer` records the history of the corresponding scalar such as losses, learning rates, and iteration time in an array and as an internal class, works with [MessageHub](mmengine.logging.MessageHub), LoggerHook and [LogProcessor](mmengine.runner.LogProcessor) to make training log configurable. Meanwhile, HistoryBuffer can also be used alone, which enables users to manage their training logs and do various statistics in an easy manner.
We will first introduce the usage of HistoryBuffer in the following section. The association between HistoryBuffer and MessageHub will be introduced later in the MessageHub section.
```

> ### Initialize HistoryBuffer
> 
> HistoryBuffer accepts 2 arguments, named `log_history` and `count_history`, for initialization. `log_history` records the trajectory of the training. For example, if the losses in the previous 3 iterations are 0.3, 0.2, 0.1 respectively, then we have `log_history=[0.3, 0.2, 0.1]`. `count_history` is more indirect. It controls the statistical granularity and is only applied when counting the average. Take the above example, if we count the average loss across iterations, then we have `count_history=[1, 1, 1]`. Instead, if we count the average loss across images with `batch_size=8`, then we have `count_history=[8, 8, 8]`.

Reviewer: ```suggestion
### HistoryBuffer Initialization

HistoryBuffer accepts `log_history`, `count_history` and `max_length` for initialization. 
* `log_history` records the history of the scaler. For example, if the loss in the previous 3 iterations is 0.3, 0.2, 0.1 respectively, the `log_history` will be [0.3, 0.2, 0.1]. 
* `count_history` controls the statistical granularity and will be used when counting the average. Take the above example, if we count the average loss across iterations, we have `count_history=[1, 1, 1]`. Instead, if we count the average loss across images with `batch_size=8`, then we have `count_history=[8, 8, 8]`.
* `max_length`controls the maximum length of the history. If the length of `log_history` and `count_history` exceeds `max_length`, the earliest elements will be removed.
Besides, We can access the trajectory of the data through `history_buffer.data`.
```

```python
from mmengine.logging import HistoryBuffer

history_buffer = HistoryBuffer()  # Default initialization
log_history, count_history = history_buffer.data
# [] []
history_buffer = HistoryBuffer([1, 2, 3], [1, 2, 3])  # Init with lists
log_history, count_history = history_buffer.data
# [1 2 3] [1 2 3]
history_buffer = HistoryBuffer([1, 2, 3], [1, 2, 3], max_length=2)
# The length of history buffer(3) exceeds the max_length(2), the first few elements will be ignored.
log_history, count_history = history_buffer.data
# [2 3] [2 3]
```

> We can access the trajectory of the data through `history_buffer.data`. Besides, we can also set the maximum length of the HistoryBuffer. When the length of a HistoryBuffer exceeds its maximum length, the earliest few data items will be dropped.

Reviewer: ```suggestion
```

> ### Update HistoryBuffer
> 
> We can update the HistoryBuffer with `update` method. It accepts 2 arguments, where the first updates `log_history` while the second updates `count_history`.

Reviewer: ```suggestion
### HistoryBuffer Update

We can update the `log_history` and `count_history` through `HistoryBuffer.`update(log_history, count_history)`.
```

```python
history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
history_buffer.update(4)  # count default to 1
log_history, count_history = history_buffer.data
# [1, 2, 3, 4] [1, 1, 1, 1]
history_buffer.update(5, 2)
log_history, count_history = history_buffer.data
# [1, 2, 3, 4, 5] [1, 1, 1, 1, 2]
```

> ### Basic statistical methods

Reviewer: ```suggestion
### Basic Statistical Methods
```

HistoryBuffer provides some basic statistical methods:

- `current()`: Get the latest data.
- `mean(window_size=None)`: Count the mean value of the previous `window_size` data. Defaults to None, as global mean.
- `max(window_size=None)`: Count the max value of the previous `window_size` data. Defaults to None, as global maximum.
- `min(window_size=None)`: Count the min value of the previous `window_size` data. Defaults to None, as global minimum.

```python
history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
history_buffer.min(2)
# 2, the minimum in [2, 3]
history_buffer.min()
# 1, the global minimum

history_buffer.max(2)
# 3，the maximum in [2, 3]
history_buffer.min()
# 3, the global maximum
history_buffer.mean(2)
# 2.5，the mean value in [2, 3], (2 + 3) / (1 + 1)
history_buffer.mean()
# 2, the global mean, (1 + 2 + 3) / (1 + 1 + 1)
history_buffer = HistoryBuffer([1, 2, 3], [2, 2, 2])  # Cases when counts are not 1
history_buffer.mean()
# 1, (1 + 2 + 3) / (2 + 2 + 2)
history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
history_buffer.update(4, 1)
history_buffer.current()
# 4
```

> ### Register statistical methods
> 
> In order to make HistoryBuffer extensible, we provide `register_statistics` method for users to register their custom statistical methods.
> 
> ```python
> from mmengine.logging import HistoryBuffer
> import numpy as np
> 
> 
> @HistoryBuffer.register_statistics
> def weighted_mean(self, window_size, weight):
>     assert len(weight) == window_size
>     return (self._log_history[-window_size:] * np.array(weight)).sum() / \
>             self._count_history[-window_size:]
> 
> 
> history_buffer = HistoryBuffer([1, 2], [1, 1])
> history_buffer.statistics('weighted_mean', 2, [2, 1])  # get (2 * 1 + 1 * 2) / (1 + 1)
> ```
> 
> You can call `statistics` to use registered statistical method by providing method name and extra arguments, as explained below.
> 
> ### Entry point of statistical methods
> 
> We provide a method, signatured as `statistics(name, *args, **kwargs)`, in HistoryBuffer to call any statistical method with any number of arguments for extensibility. The `name` parameter should be a registered method name (i.e. built-in methods like `min` and `max`), while `*args` and `**kwargs` should be corresponding method's arguments.
> 
> ```python
> history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
> history_buffer.statistics('mean')
> # 2, as global mean
> history_buffer.statistics('mean', 2)
> # 2.5, as the mean of [2, 3]
> history_buffer.statistics('mean', 2, 3)
> # Error! mismatch arguments given to `mean(window_size)`
> history_buffer.statistics('data')
> # Error! `data` method not registered
> ```

Reviewer: ```suggestion
### Statistical Methods Invoking

Statistical methods can be accessed through `HistoryBuffer.statistics` with method name and arguments. The `name` parameter should be a registered method name (i.e. built-in methods like `min` and `max`), while arguments should be the corresponding method's arguments.

\```python
history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
history_buffer.statistics('mean')
# 2, as global mean
history_buffer.statistics('mean', 2)
# 2.5, as the mean of [2, 3]
history_buffer.statistics('mean', 2, 3)
# Error! mismatch arguments given to `mean(window_size)`
history_buffer.statistics('data')
# Error! `data` method not registered
\```

### Statistical Methods Registration

Custom statistical methods can be registered through `@HistoryBuffer.register_statistics`.

\```python
from mmengine.logging import HistoryBuffer
import numpy as np

@HistoryBuffer.register_statistics
def weighted_mean(self, window_size, weight):
    assert len(weight) == window_size
    return (self._log_history[-window_size:] * np.array(weight)).sum() / \
            self._count_history[-window_size:]

history_buffer = HistoryBuffer([1, 2], [1, 1])
history_buffer.statistics('weighted_mean', 2, [2, 1])  # get (2 * 1 + 1 * 2) / (1 + 1)
\```

```

Author: Mark, Chinese docs should also be changed

> ### Use cases

Reviewer: ```suggestion
### Use Cases
```

                            'train/data_time': {'value': 0.1, 'count': 1}})

train_time = message_hub.get_scalar('train/time')  # 1

log_dict = message_hub.log_scalars  # return the whole dict
lr_buffer, loss_buffer, time_buffer, data_time_buffer = (
    log_dict['train/lr'], log_dict['train/loss'], log_dict['train/time'],
    log_dict['train/data_time'])
```

```{note}
Losses, learning rates and iteration time are automatically updated by runner and hooks. You are not supposed to manually update them.
```

```{note}
MessageHub has no special requirements for keys in `log_scalars`. However, MMEngine will only output a scalar to logs if it has a key prfixed with train/val/test.
```

### Update & get runtime info

Runtime information is stored in `runtime_info` dict. The dict accepts data in any data types. Different from HistoryBuffer, the value will be overwritten on every update.

```python
message_hub = MessageHub.get_instance('task')
message_hub.update_info('iter', 1)
message_hub.get_info('iter')  # 1
message_hub.update_info('iter', 2)
message_hub.get_info('iter')  # 2, overwritten by the above command
```

### Share MessageHub across modules

> During the execution of a runner, different modules receive and post data through MessageHub. Then, [RuntimeInfoHook](mmengine.hooks.RuntimeInfoHook) gathers data such as losses and learning rates before exporting them to user defined backends(Tensorboard, WandB, etc). Following is an example to show the communication between logger hook and other modules.

Reviewer 2: ```suggestion
During the execution of a runner, different modules receive and post data through MessageHub. Then, [RuntimeInfoHook](mmengine.hooks.RuntimeInfoHook) gathers data such as losses and learning rates before exporting them to user defined backends (Tensorboard, WandB, etc). Following is an example to show the communication between logger hook and other modules.
```

- `method_name`: Required argument representing the statistical method. It supports both built-in methods and custom methods.
- `log_name`: Optional argument representing the output name after statistics. If not specified, the new log will overwrite the old one.
- Other arguments: Extra arguments needed by your specified method. `window_size` is a special key, which can be either an int, 'epoch' or 'global'. LogProcessor will parse these arguments and return statistical result based on iteration/epoch/global smooth.

1. Overwrite the old statistical method

```python
log_processor = dict(
    window_size=10,
    by_epoch=True,
    custom_cfg=[
        dict(data_src='loss',
             method_name='mean',
             window_size=100)])
```

> In this configuration, LogProcessor will overwrite the default window size 10 by a larger window size 100 and output the mean value to 'loss' field in logs.

Reviewer 2: ```suggestion
In this configuration, LogProcessor will overwrite the default window size 10 by a larger window size 100 and output the mean value to 'loss' field in logs.
```
